sycl/source/detail/graph_impl.cpp

```diff
@@ -441,69 +509,18 @@ graph_impl::add(const std::vector<sycl::detail::EventImplPtr> Events) {
 graph_impl::add(node_type NodeType,
-                std::unique_ptr<sycl::detail::CG> CommandGroup,
+                std::shared_ptr<sycl::detail::CG> CommandGroup,
                 const std::vector<std::shared_ptr<node_impl>> &Dep) {
-  // Copy deps so we can modify them
-  auto Deps = Dep;
 
   // A unique set of dependencies obtained by checking requirements and events
-  std::set<std::shared_ptr<node_impl>> UniqueDeps;
-  const auto &Requirements = CommandGroup->getRequirements();
-  if (!MAllowBuffers && Requirements.size()) {
-    throw sycl::exception(make_error_code(errc::invalid),
-                          "Cannot use buffers in a graph without passing the "
-                          "assume_buffer_outlives_graph property on "
-                          "Graph construction.");
-  }
+  std::set<std::shared_ptr<node_impl>> UniqueDeps = getCGEdges(CommandGroup);
 
-  if (CommandGroup->getType() == sycl::detail::CGType::Kernel) {
-    auto CGKernel =
-        static_cast<sycl::detail::CGExecKernel *>(CommandGroup.get());
-    if (CGKernel->hasStreams()) {
-      throw sycl::exception(
-          make_error_code(errc::invalid),
-          "Using sycl streams in a graph node is unsupported.");
-    }
-  }
-
-  for (auto &Req : Requirements) {
-    // Track and mark the memory objects being used by the graph.
-    auto MemObj = static_cast<sycl::detail::SYCLMemObjT *>(Req->MSYCLMemObj);
-    bool WasInserted = MMemObjs.insert(MemObj).second;
-    if (WasInserted) {
-      MemObj->markBeingUsedInGraph();
-    }
-    // Look through the graph for nodes which share this requirement
-    for (auto &Node : MNodeStorage) {
-      if (Node->hasRequirementDependency(Req)) {
-        bool ShouldAddDep = true;
-        // If any of this node's successors have this requirement then we skip
-        // adding the current node as a dependency.
-        for (auto &Succ : Node->MSuccessors) {
-          if (Succ.lock()->hasRequirementDependency(Req)) {
-            ShouldAddDep = false;
-            break;
-          }
-        }
-        if (ShouldAddDep) {
-          UniqueDeps.insert(Node);
-        }
-      }
-    }
-  }
+  // Track and mark the memory objects being used by the graph.
+  markCGMemObjs(CommandGroup);
 
-  // Add any nodes specified by event dependencies into the dependency list
-  for (auto &Dep : CommandGroup->getEvents()) {
-    if (auto NodeImpl = MEventsMap.find(Dep); NodeImpl != MEventsMap.end()) {
-      UniqueDeps.insert(NodeImpl->second);
-    } else {
-      throw sycl::exception(sycl::make_error_code(errc::invalid),
-                            "Event dependency from handler::depends_on does "
-                            "not correspond to a node within the graph");
-    }
-  }
-  // Add any deps determined from requirements and events into the dependency
-  // list
+  // Copy deps so we can modify them. Add any deps determined from requirements
+  // and events into the dependency list
+  auto Deps = Dep;
```

Comment:
`const_cast` always feels a bit like a workaround, but agree this copy is awkward. I've changed the way this parameter is passed so it's a reference rather than a const reference and we can modify it here without the copy.